readme.md
# NDSA - Nintendo DS Advanced Engine
## Write for NDS and GBA in one codebase!
A project from 2016 that I'd like to bring back. It's a header library which wraps libnds/libgba functions, adds an object system, so on. NOTE: Requires [gritn](https://github.com/nicklausw/gritn). Other than that, if you have devkitPro installed, you should be able to clone the repo, run GNU make and have fun.

## Objects
Declare an object like so:
```c++
struct Player : Object {
  using ParentConstructors;
  void Step() override {
    // code run once per frame
  }
};
```
Run it like so:
```c++
// this object uses the sprite manSprite at (150, 50)
Player *player = new Player(manSprite, 150, 50, TopScreen);
```
Don't need a sprite? Don't give the object one.
```c++
Player *player = new Player();
```

## Sprites
For now, use 8-bit paletted images. See [example_code/images](example_code/images) directory for examples of proper `.grit` files to go with them.
```c++
includeTiles(man);
Sprite *manSprite = new Sprite(TileData(man), Size_32x32, Colors256);
```
### Spritesheets
```c++
includeTiles(spriteSheet);
SpriteSheet *spriteSheet = new SpriteSheet(TileData(spriteSheet), Size_32x32, Colors256);
Sprite *manSprite = spriteSheet->getByIndex(3);
```

## Input
```c++
if(Buttons.A.Pressed()) {
  // do something...
} else if(Buttons.Up.Held()) {
  // do something else...
}
```

## Backgrounds
Same deal as sprites; use 8-bit paletted images.
```c++
includeTiles(bg);
includeMap(bg); // this can be from bg.png or bg.tmx (tilemap)
Background.Set(TileData(bg), MapData(bg), TopScreen);
```

## Random Numbers
```c++
// get random number
return Random.Next();
```

## The Main Function
Use this in favor of your own `int main`:
```c++
#include <NDSA/Main.hh>
void NDSA::Game() {
  while(System.Frame()) {
    // this code runs once every frame
  }
}
```

changelog.md
### 6/25/16 - v1.0
* the initial release.
* button system, only pressed so far.
* pretty frames (with vblank and buttons!).

### 6/26/16 - v1.1
* gave NDSA_Keys a struct home for multi-module dev.
* added button Held() function.
* added clean touch screen support.

### 6/29/16 - v1.2
* added random numbers, Mersenne Twister and C-style.
* added not-quite-working sprites
* added basically working objects?

### 7/3/16 - v1.3
* added properly working sprites
* fixed objects
* added basic audio playback
* DS.Initialize() is a thing, remove if possible

### 7/6/16 - v1.4
* made audio optional
* added backgrounds
* added dual-screen support!
* todo: documentation.....:(

### 1/16/22 - v1.5
* welcome to the next decade!
* updated to repo format
* MIT license
* plenty of code improvements

### x/xx/xx - v2.0?
* GBA support is here!
* large tiled background support
* sprite sheet support
* use c functions rather than std
* PointerList!!!
* objects can be accessed by type/index

example_code/multi_module_test.cpp
#include <NDSA.hh>
using namespace NDSA;

void printFrame(int frame) {
  PrintAt(0, 10, "Frame = %d", frame);
}

tools/mapToC.py
#!/usr/bin/env python3
import sys
import os
if len(sys.argv) != 4:
  print("usage: " + sys.argv[0] + " (csv file) (c output) (h output)")
  exit(1)
count = 0
head, tail = os.path.split(sys.argv[1])
variableName = tail.removesuffix(".csv") + "Data"

data = ""
with open(sys.argv[1], 'r') as fd:
  lines = fd.readlines()
  for line in lines:
    count += line.count(",") + 1
    data += "  " + line.rstrip("\n") + ",\n"

h_output = "#ifndef " + variableName + "_h\n" + \
  "  #define " + variableName + "_h\n" + \
  "  extern const short unsigned int " + variableName + "[" + str(count) + "];\n" + \
  "  #define " + variableName + "Length " + str(count * 2) + "\n" \
  "#endif"
output = "const short unsigned int " + variableName + "[" + str(count) + "] = {\n"
output += data
output = output.rstrip(",\n") + "\n};\n"

text_file = open(sys.argv[2], "w")
text_file.write(output)
text_file.close()

text_file = open(sys.argv[3], "w")
text_file.write(h_output)
text_file.close()

tools/mapToS.py
#!/usr/bin/env python3
import sys
import os
import xml.etree.ElementTree as ET
if len(sys.argv) != 3:
  print("usage: " + sys.argv[0] + " (tmx file) (s output)")
  sys.exit(1)
count = 0
head, tail = os.path.split(sys.argv[1])
variableName = tail.removesuffix(".tmx")

rawData = []
smallestInt = 100000
xmlDat = ET.parse(sys.argv[1]).getroot()

rawWidth = int(xmlDat.attrib["width"])
rawHeight = int(xmlDat.attrib["height"])
if rawWidth % 32 != 0:
  sys.stderr.write("error: " + tail + ": width of map must be divisible by 32\n")
  sys.exit(1)
if rawHeight % 32 != 0:
  sys.stderr.write("error: " + tail + ": height of map must be divisible by 32\n")
  sys.exit(1)

rawText = xmlDat.find("layer").find("data").text
lines = rawText.split("\n")
for line in lines:
  if(len(line)):
    nums = line.split(",")
    for num in nums:
      if len(num):
        count += 1
        newNum = int(num)
        rawData.append(newNum)
        if newNum < smallestInt:
          smallestInt = newNum

output = (
  "@ generated by mapToS.py\n"
  "@ data size = " + str(count * 2 + 4) + " bytes\n\n"
  ".section .rodata\n"
  ".align 4\n\n"
  ".global " + variableName + "MapLen\n"
  ".hidden " + variableName + "MapLen\n"
  + variableName + "MapLen:\n"
  ".long " + "0x{:08X}".format(count * 2) + "\n\n"
  ".global " + variableName + "Map\n"
  ".hidden " + variableName + "Map\n" + variableName + "Map:\n"
)

listWidth = int(rawWidth / 32)
listHeight = int(rawHeight / 32)

columnBase = 0
rowBase = 0
outData = []

for row in range(0, listHeight):
  rowBase = 0
  for column in range(0, listWidth):
    for x in range(0,32):
      for y in range(0,32):
        outData.append(rawData[columnBase + (x * listWidth * 32) + rowBase + y])
    rowBase += 32
  columnBase += (listWidth * 32 * 32) 

counter = 0
for num in outData:
  if counter == 0:
    output += ".hword "
  output += "0x{:04X}".format(num - smallestInt)
  counter += 1
  if counter == 8:
    output += "\n"
    counter = 0
  else:
    output += ","
output += "\n"

text_file = open(sys.argv[2], "w")
text_file.write(output)
text_file.close()

example_code/main.cpp
#include <NDSA.hh>
using namespace NDSA;

#include "testImage.h"
#include "man.h"
#include "enemy.h"

#define SCREEN(x,y) "\x1b["#y";"#x"H"

struct Custom : Object {
  int frame;
  Custom() {
    frame = 0;
  }
  void Run() override {
    if(!TouchScreen.Held()) {
      frame++;
    } else {
      frame -= 2;
    }
    std::cout << SCREEN(0,10) << "Frame = " << frame;
  }
};

struct Player : Object {
  using ParentConstructors;
  ~Player() {
    std::cout << SCREEN(0,6) << "You killed him! Touch again to  "
                                "make another.";
  }
  void Run() override {
    if(Buttons.Up.Held()) {
      Move(Up, .1);
    } else if(Buttons.Down.Held()) {
      Move(Down, .1);
    }
    if(Buttons.Left.Held()) {
      Move(Left, .1);
    } else if(Buttons.Right.Held()) {
      Move(Right, .1);
    }
  }
};

struct Enemy : Object {
  using ParentConstructors;
};

#include <NDSA/Main.hh>
void NDSA::Game() {
  Custom *custom = new Custom();
  Sprite *manSprite = new Sprite(SpriteData(man), SpriteSize_32x32, SpriteColorFormat_256Color);
  Sprite *enemySprite = new Sprite(SpriteData(enemy), SpriteSize_32x32, SpriteColorFormat_256Color);
  Player *player = new Player(manSprite, 150, 50, TopScreen);
  Enemy *enemy = new Enemy(enemySprite, 50, 50, TopScreen);


  consoleDemoInit();
  Background.Set(BackgroundData(testImage), TopScreen);

  std::cout << SCREEN(0,3) << "NDSA is awesome!";
  std::cout << SCREEN(0,4) << "Move the guy with the D-pad.";
  std::cout << SCREEN(0,6) << "Touch the screen to delete him.";

  bool deleted = false;
  while(DS.Frame()) {
    if(TouchScreen.Tapped()) {
      if(deleted == false) {
        deleted = true;
        delete player;
      } else {
        deleted = false;
        player = new Player(manSprite, 150, 50, TopScreen);
        std::cout << SCREEN(0,6) << "Hey, he's back!                 "
                                    "                                ";
      }
    }
  };
}

example_code/platformer.cpp
#include <NDSA.hh>
using namespace NDSA;

enum Status {
  idle,
  falling,
  jumping
};

double movement(double velocity, double acceleration, int step) {
  return ((velocity * step) + ((1.0/2.0) * (acceleration) * pow(step,2.0))) * .001;
}

double movementROC(double velocity, double acceleration, int step) {
  return (velocity + (2.0 * acceleration * step)) * .001;
}

struct Bullet : Object {
  using ParentConstructors;
  void Step() override {
    Move(Right, 10);
    if(X >= SCREEN_WIDTH - xDeadSpace() - width) {
      delete this;
    }
  }
};

struct Enemy : Object {
  using ParentConstructors;
  void Step() override {
    Move(Left, .5);
    if(X - xDeadSpace() <= 0.0F) {
      delete this;
    }
  }
};

struct Player : Object {
  using ParentConstructors;

  Status yStatus = falling;

  int health = 10;
  int invincFrame = 0;

  float yROC = 0; // y Rate of Change
  int yStep = 0;
  int xStep = 0;
  const float xVelocity = 100;
  const float yVelocity = 1.0;
  const float xAccel = .1;
  const float xMaxStep = 200;
  float yBase = 0;

  Sprite *bulletSpritePointer;

  void Step() override {
    if(invincFrame) invincFrame--;
    if(Buttons.A.Pressed() && yStatus == idle) {
      yStatus = jumping;
      yROC = 1;
      yBase = Y;
      yStep = 0;
    }
    if(Buttons.B.Pressed()) {
      Bullet *b = new Bullet(bulletSpritePointer, round(X + 8), round(Y)
                             #ifdef DS
                             , TopScreen
                             #endif
                             );
      b->addCollision(8, 8);
    }
    if(Buttons.Left.Held()) {
      if(xStep > -xMaxStep) xStep-=2;
      Move(Right, movement(xVelocity, (xStep > 0 ? xAccel : -xAccel), xStep));
    } else if(Buttons.Right.Held()) {
      if(xStep < xMaxStep) xStep+=2;
      Move(Right, movement(xVelocity, (xStep > 0 ? xAccel : -xAccel), xStep));
    } else if(xStep) {
      Move(Right, movement(xVelocity, (xStep > 0 ? xAccel : -xAccel), xStep));
      if(xStep > 0) xStep-=2; else xStep+=2;
    }
    if(yStatus == jumping) {
      Y = yBase - movement(yVelocity, -9.8, yStep);
      yStep+=2;
      if(Y > yBase) {
        Y = yBase;
        yStatus = idle;
        isFalling = false;
      }
      Update();
    }
    else if(yStatus == falling) {
      yROC = movement(yVelocity, -9.8, yStep);
      Move(Up, movement(yVelocity, -9.8, yStep));
      yStep++;
    }
  }
};

includeTiles(bgTileSet);
includeTiles(spriteSheet);
includeMap(tileMap);

void onEnemyBulletCollide(Enemy *p, Bullet *b) {
  Objects.deleteByInstance<Enemy>(&p);
  Objects.deleteByInstance<Bullet>(&b);
}

void onPlayerEnemyCollide(Player *p, Enemy *e) {
  if(!p->invincFrame) {
    p->health--;
    p->invincFrame = 500;
  }
}

#include <NDSA/Main.hh>
void NDSA::Game() {
  SpriteSheet *spriteSheet = new SpriteSheet(TileData(spriteSheet), Size_32x32, Colors256);
  Sprite *manSprite = spriteSheet->getByIndex(3);
  Sprite *enemySprite = spriteSheet->getByIndex(1);
  Sprite *bulletSprite = spriteSheet->getByIndex(2);

  Background.Set(TileData(bgTileSet), MapData(tileMap), BGSize_512x512
                 #ifdef DS
                 , TopScreen
                 #endif
                 );

  Player *player = new Player(manSprite, 50, 50
                              #ifdef DS
                              , TopScreen
                              #endif
                              );
  player->bulletSpritePointer = bulletSprite;
  player->addCollision(32, 32);

  onCollision<Enemy, Bullet>(onEnemyBulletCollide);
  onCollision<Player, Enemy>(onPlayerEnemyCollide);
  
  #ifdef DS
  initConsole();
  #endif
  
  Background.setScrollY(512 - SCREEN_HEIGHT);
  #ifdef DS
	bgUpdate();
  #endif

  int frame=0;
  while(System.Frame()) {
    frame++;
    #ifdef DS
    PrintAt(0,5,"Player health: %d    ",player->health);
    #endif
    if(frame == 250) {
      Enemy *e = new Enemy(enemySprite, SCREEN_WIDTH, SCREEN_HEIGHT - 64
                           #ifdef DS
                           , TopScreen
                           #endif
                           );
      e->addCollision(24, 32);
      frame = 0;
    }
    Object *o = player;
    const MapData* tileMapData = tileMapMap;
    float collisionXtop = (o->X / 8.0) - (Background.getScrollX() / 8.0);
    float collisionYtop = (o->Y / 8.0) + (Background.getScrollY() / 8.0);
    float collisionXbottom = ((o->X + o->width) / 8.0) - (Background.getScrollX() / 8.0);
    float collisionYbottom = ((o->Y + o->height) / 8.0) + (Background.getScrollY() / 8.0);
    int collisionNumTop = tileMapData[((int)floor(collisionYtop) * (512 / 8)) + (int)collisionXtop];
    int collisionNumBottom = tileMapData[((int)floor(collisionYbottom) * (512 / 8)) + (int)collisionXbottom];
    if((collisionNumTop || collisionNumBottom) && player->yROC < 0) {
      player->Y = floor(player->Y / 8) * 8;
      player->Update();
      player->yStatus = idle;
    }
    #ifdef DS
    PrintAt(0,2,"# of objects: %d    ",Objects.count);
    #endif
  }
}

example_code/example.cpp
#include <NDSA.hh>
using namespace NDSA;

#include "testImage.h"
#include "man.h"
#include "enemy.h"
#include "bullet.h"

// ensure multi-module support
void printFrame(int);

struct Custom : Object {
  int frame;
  Custom() {
    frame = 0;
  }
  void Step() override {
    if(!TouchScreen.Held()) {
      frame++;
    } else {
      frame -= 2;
    }
    printFrame(frame);
  }
};

struct Player : Object {
  using ParentConstructors;
  ~Player() {
    PrintAt(0, 6, "You killed him! Touch again to  "
                  "make another.");
  }
  void Step() override {
    if(Buttons.Up.Held()) {
      Move(Up, .2);
    } else if(Buttons.Down.Held()) {
      Move(Down, .2);
    }
    if(Buttons.Left.Held()) {
      Move(Left, .2);
    } else if(Buttons.Right.Held()) {
      Move(Right, .2);
    }
  }
};

struct Bullet : Object {
  using ParentConstructors;
  void Step() override {
    Move(Right, .1);
    if(X > 220.0F) {
      delete this;
    }
  }
};

struct Enemy : Object {
  using ParentConstructors;
  int frame = 0;
  Sprite *bulletSpritePointer;
  Direction d = Down;
  void Step() override {
    frame++;
    if(frame == 50) {
      Bullet *newB = new Bullet(bulletSpritePointer, round(X + 8), round(Y + 16), TopScreen);
      frame = 0;
    }

    Move(d, .075);
    if(Y < 20) {
      d = Down;
    } else if(Y > 150) {
      d = Up;
    }
  }
};

#include <NDSA/Main.hh>
void NDSA::Game() {
  Sprite *manSprite = new Sprite(SpriteData(man), SpriteSize_32x32, SpriteColorFormat_256Color);
  Sprite *enemySprite = new Sprite(SpriteData(enemy), SpriteSize_32x32, SpriteColorFormat_256Color);
  Sprite *bulletSprite = new Sprite(SpriteData(bullet), SpriteSize_8x8, SpriteColorFormat_256Color);

  Custom *custom = new Custom();
  Player *player = new Player(manSprite, 150, 50, TopScreen);
  Enemy *enemy = new Enemy(enemySprite, 50, 50, TopScreen);
  enemy->bulletSpritePointer = bulletSprite;


  consoleDemoInit();
  Background.Set(BackgroundData(testImage), TopScreen);

  PrintAt(0, 3, "NDSA is awesome!");
  PrintAt(0, 4, "Move the guy with the D-pad.");
  PrintAt(0, 6, "Touch the screen to delete him.");
  
  PointerList<Object> *oList = new PointerList<Object>();
  while(DS.Frame()) {
    if(Buttons.A.Pressed()) {
      Objects.deleteByType<Bullet*>();
    }
    if(Buttons.B.Held()) {
      Object *o = new Object();
      Object *p = new Object();
      Object *z = new Object();
      oList->add(o);
      oList->add(p);
      oList->add(z);
    }
    if(Buttons.Y.Pressed()) {
      oList->deleteAll();
    }
    if(TouchScreen.Tapped()) {
      if(player) {
        Objects.deleteByInstance<Player>(&player);
      } else {
        player = new Player(manSprite, 150, 50, TopScreen);
        PrintAt(0, 6, "Hey, he's back!                 "
                      "                                ");
      }
    }
    int count = Objects.count, size = Objects.listSize;
    PrintAt(0, 12, "Number of objects: %d    ", count);
    PrintAt(0, 14, "Objects list size: %d    ", size);
    PrintAt(0, 16, "Random number: %ld", Random.Next());
    PrintAt(0, 18, "Number of bullets: %d    ", Objects.getByType<Bullet>().count);
  };
}